Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,267 @@
+  case prec f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.prec f' g' g ih1 ih2
+  case rfind f' _ ih =>
+    apply RecursiveIn.rfind f' g ih
+
+/--
+If a function is recursive in the constant zero function,
+then it is partial recursive.
+-/
+lemma RecursiveIn.partrec_of_zero (f : ℕ →. ℕ) (fRecInZero : RecursiveIn f fun _ => Part.some 0) :
+  Nat.Partrec f := by
+  generalize h : (fun _ => Part.some 0) = fp at *
+  induction fRecInZero
+  case zero =>
+    apply Nat.Partrec.zero
+  case succ =>
+    apply Nat.Partrec.succ
+  case left =>
+    apply Nat.Partrec.left
+  case right =>
+    apply Nat.Partrec.right
+  case oracle g =>
+    rw [← h]
+    apply Nat.Partrec.zero
+  case pair _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.pair
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case comp _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.comp
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case prec _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.prec
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case rfind _ _ ih =>
+    apply Nat.Partrec.rfind
+    apply ih
+    rw [← h]
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything (f : ℕ →. ℕ) : Nat.Partrec f ↔ ∀ g, RecursiveIn f g :=
+  ⟨(·.recursiveIn), (· _ |>.partrec_of_zero)⟩
+
+/--
+Proof that turing reducibility is reflexive.
+-/
+theorem RecursiveIn.refl (f : ℕ →. ℕ) : RecursiveIn f f :=
+  RecursiveIn.oracle f
+
+/--
+Instance declaring that `RecursiveIn` is reflexive.
+-/
+instance : IsRefl (ℕ →. ℕ) RecursiveIn :=
+  ⟨fun f => RecursiveIn.refl f⟩
+
+/--
+Proof that turing reducibility is transitive.
+-/
+theorem RecursiveIn.trans {f g h : ℕ →. ℕ} (hg : RecursiveIn f g) (hh : RecursiveIn g h) :
+  RecursiveIn f h := by
+  induction hg
+  case zero =>
+    apply RecursiveIn.zero
+  case succ =>
+    apply RecursiveIn.succ
+  case left =>
+    apply RecursiveIn.left
+  case right =>
+    apply RecursiveIn.right
+  case oracle =>
+    exact hh
+  case pair f' h' _ _ hf_ih hh_ih =>
+    apply RecursiveIn.pair
+    · apply hf_ih
+      apply hh
+    · apply hh_ih
+      apply hh
+  case comp f' h' _ _ hf_ih hh_ih =>
+    apply RecursiveIn.comp
+    · apply hf_ih
+      apply hh
+    · apply hh_ih
+      apply hh
+  case prec f' h' _ _ hf_ih hh_ih =>
+    apply RecursiveIn.prec
+    · apply hf_ih
+      apply hh
+    · apply hh_ih
+      apply hh
+  case rfind f' _ hf_ih =>
+    apply RecursiveIn.rfind
+    · apply hf_ih
+      apply hh
+
+/--
+Instance declaring that `RecursiveIn` is transitive.
+-/
+instance : IsTrans (ℕ →. ℕ) RecursiveIn :=
+  ⟨@RecursiveIn.trans⟩
+
+/--
+Instance declaring that `RecursiveIn` is a preorder.
+-/
+instance : IsPreorder (ℕ →. ℕ) RecursiveIn where
+  refl := RecursiveIn.refl
+  trans := @RecursiveIn.trans
+
+/--
+Proof that `TuringEquivalent` is an equivalence relation.
+-/
```

Comment:
Generally we prefer to omit docstrings that are a literal symbol reading of the lean code. This analogous to writing `i++; // add one to i` in C++.